Adds Embedded Cluster lab (#346)
replicatedhq/replicated-field-labs: instruqt/delivering-as-an-appliance/02-customize-for-your-app/assignment.md

---
slug: customize-for-your-app
id: 6003pttnmqok
type: challenge
title: Customizing the Appliance for Your Application
teaser: Make the Kubernetes appliance experience your own
notes:
- type: text
  contents: Customize the Admin Console to represent your application
tabs:
- id: ybi0oubv14zx
  title: Release Editor
  type: code
  hostname: shell
  path: /home/replicant
- id: qgpmbdncfnse
  title: Shell
  type: terminal
  hostname: shell
difficulty: basic
timelimit: 300
---

The Kubernetes appliance installation is one of the first things
your customer will experience about your application. The way we've implemented
it so far uses a lot of defaults. We didn't try to install it yet, but if we
had, we'd land on a page like this:

![Appliance login page without branding](../assets/unbranded-admin-console.png)

This definitely violates our brand standards!

Don't worry, Replicated provides a few options to customize the appliance

> installer. The first customizations we're going to look at will make it clear

Reviewer: "to customize the appliance installer" --> "customize the Admin Console" (maybe just to be very specific and match the title of the section?)

Author: I split the difference 😉

which application we're installing.

Adding a Name and Logo to the Appliance
=======================================

We've already looked at two of the resources we use for configuring the
appliance experience: one for the embedded Kubernetes cluster and one for the
SlackerNews Helm chart. There are others as well. We're going to look at the
`Application` resource, which describes the application being installed. There
are several things we can configure in this file as well, but for this lab

> we'll focus n the title, icon, and how to determine application status.. For

Reviewer: ```suggestion
we'll focus on the title, icon, and how to determine application status. For
```

other possible properties, take a look at the [reference
documentation](https://docs.replicated.com/reference/custom-resource-application).
Not all of the possible values apply for the Kubernetes appliance experience.

The appliance installation process also supports another resource that
describes application metadata. The resource is defined by the
Kubernetes Apps Special Interest Group (SIG) and is also called `Application`.
This `Application` resource is not unique to the Replicated Platform. It's
also not required for the appliance installation experience.

> We'll start with adding the basic Replicated `Application` resource for basic

Reviewer: ```suggestion
We'll start with adding the Replicated `Application` resource for basic
```

Just so that basic's not repeated twice

branding, then we'll expand on it a bit. We're going to skip the SIG
`Application` for this lab.

### Add the Application Manifest

Let's go back to the "Release Editor" tab and add a new file for the
application metadata. Create a new file in the `release` directory named
`replicated-app.yaml`
it.

![Adding the application metadata manifest](../assets/adding-application-manifest.png)

> Let's start with just the name and log fro the application. This will make

Reviewer: ```suggestion
Let's start with just the name and log for the application. This will make
```

needed load balancers have been assigned. They are degraded if at least one
endpoint is ready when more are requests, and unavailable if none are ready
and no load balancer(s) have been assigned. Ingress is based on the same
criteria for it's backing service.</td></tr>
<tr><td><code>PersistentVolmeClaim</code></td><td>A PVC is ready if the claim is bound and
unavailable if it is either lost or pending. Those are the only two states.</td></tr>
</tbody>
<table>

SlackerNews has three components in it's architecture: the application, and
NGINX web server, and a Postgres database. It can optionally have an ingress
as well, but will not in the virtual appliance. Let's add status informers for
the application and the web server.

In the "Release Editor" tab, add the following to the `spec` section of the
manifest.

```
  statusInformers:
    - deployment/slacknerews
    - deployment/slackernews-nginx
```

![Adding the application metadata manifest](../assets/adding-status-informers.png)

We haven't provided a status informer for Postgres since the customer can
choose to bring their own database. We'll add it in when we work with the
customer configuration in the next step of this lab.

> Releasing Your Application

Reviewer: ```suggestion
Releasing the Appliance
```
To match how we're talking about it elsewhere

==========================

We're still not going to install our appliance, but let's release these
changes to get into the habit of releasing changes quickly.

First we want to bump the version in our Helm chart. You could consider this
part optional since we didn't update the chart. I recommend you do it, since
different customer may install the application in different ways. It will be
much easier for your team if Helm installs using version `0.6.1` and appliance
installs using version `0.6.1` are using the same bits.

You can bump the chart version in your editor, or just run the following two
commands

```
yq -i '.version = "0.6.1"' slackernews/Chart.yaml
yq -i '.spec.chart.chartVersion = "0.6.1"' release/slackernews-chart.yaml
```

You need to edit both files to ensure that the appliance refers to the current

> chart. You also need to repackage you're Helm chart and make sure your old

Reviewer: ```suggestion
chart. You also need to repackage your Helm chart and make sure your old
```

installer. The first customizations we're going to look at will make it clear
which application we're installing.

Adding a Name and Logo to the Appliance
=======================================

We've already looked at two of the manifests we use for configuring the
appliance experience: one for the embedded Kubernetes cluster and one for the
Slackernews Helm chart. There are other manifests as well. We're going to look
at the `Application` manifest, which describes the application being installed.

```
apiVersion: kots.io/v1beta1
kind: Application
metadata:
  name: slackernews
spec:
  title: SlackerNews
  icon: "https://uploads-ssl.webflow.com/6310ad0e6a18aa1620da6ae8/6330e04f42bc6a7ba03b4725_snicon.png"
```

There are other things we can configure in this file as well, but for this lab
we'll focus just on the title and the icon. For other possible properties, take
a look at the [reference
documentation](https://docs.replicated.com/reference/custom-resource-application).
Not all of the possible values apply for the Kubernetes appliance experience.

The appliance installation process also supports another resource that

> describes application metadata. The resource is based defined by the

Reviewer 2: "based defined by" typo

Kubernetes Apps Special Interest Group (SIG) and is also called `Application`.
This `Application` resource is not unique to the Replicated Platform. It's
also not required for the appliance installation experience.

We'll start with adding the basic Replicated `Application` resource for basic
branding, then we'll expand on it a bit. We'll create a SIG Application
resource as part of that expansion.

### Add the Application Manifest

Let's go back to the "Release Editor" tab and add a new file for the
application metadata. Create a new file in the `release` directory named

> `replicated-app.yaml` and copy/paste the contents from the example above into

Reviewer 2: When I saw the copy for this file above, I thought I needed to create the file. Then I didn't have to until here, and I had to scroll back up to copy it. I think it's clearer to have the content right here when I need to copy them, and not earlier. IMO the link to the docs shows what the file looks like as you're reading through that earlier part.

application status indicator. This will help an administrator understand the
running state of the application and quickly connect to it.

### Reporting Status on the Admin Console

Application status is reported based on status informers identified in the
Replicated `Application` resource. Status informers help the Admin Console
determine the application status to display. The status is an aggregation
across the provided status informers.

Application status is similar to the status you see for a Kubernetes
resource when using the `kubectl` command. The statuses provided are based on
the resources configured as status informers.

<table>
<thead>
</tr><th>Status</th><th>Description</th></tr>
</thead>
<tbody>
<tr><td>Ready</td><td>All of the resources are ready</td></tr>
<tr><td>Unvailable</td><td>One or more of the resources are unavailable</td></tr>
<tr><td>Degraded</td><td>One or more of the resources are in a degraded state</td></tr>
<tr><td>Updating</td><td>There are multiple versions of the some resource present</td></tr>
<tr><td>Missing</td><td>The resources have not yet been created</td></tr>
</tbody>
</table>

> Status informers can reference on of several types of resource:

Reviewer 2: ```suggestion
Status informers can reference one of several types of resources:
```

* Deployments

> * Stateful sets
> * Daemon sets
> * Services
> * Persistent volume claims

Reviewer 2: ```suggestion
* StatefulSets
* DaemonSets
* Services
* PersistentVolumeClaims
```

Author: it's interesting. I deliberately didn't use the resource syntax and chose English. I'm curious why you opted the other way.

Reviewer: Eavesdropping: I like using the resource names in the docs so there's no confusion about the types of resources supported (most likely people know that Persistent volume claim = PersistentVolumeClaim, but using PersistentVolumeClaim slightly reduces the level of thinking required)

resource it references. Missing is the same for all resources: it has not yet
been created.

<table>
<tbody>
<tr><td><code>Deployment</code><br/><code>StatefulSet</code></td><td>Status is based on the number
of replicas that are ready. A resources is ready if all replicas are
available, degraded one but not all replicas are ready, and unavailable if
none are ready. If there are replicas are from different versions then the
resources is updating.</td></tr>
<tr><td><code>DaemonSet</code></td><td>Similar to </code>Deployment</code> or </code>StatfulSet</code> but based on
the daemon pods</td></tr>
<tr><td><code>Service</code><br/><code>Ingress</code></td><td>Services are ready if all endpoints are ready and any
needed load balancers have been assigned. They are degraded if at least one
endpoint is ready when more are requests, and unavailable if none are ready
and no load balancer(s) have been assigned. Ingress is based on the same
criteria for it's backing service.</td></tr>
<tr><td><code>PersistentVolmeClaim</code></td><td>A PVC is ready if the claim is bound and
unavailable if it is either lost or pending. Those are the only two states.</td></tr>
</tbody>
<table>

> Slackernews has three components in it's architecture: the application, and

Reviewer 2: ```suggestion
Slackernews has three components in it's architecture: the application, an
```

NGINX web server, and a Postgres database. It can optionally have an ingress
as well, but will not in the virtual appliance. Let's add status informers for
the application and the web server.

In the "Release Editor" tab, add the following to the `spec` section of the
manifest.

```
  statusInformers:
    - deployment/slacknerews
    - deployment/slackernews-nginx
```

![Adding the application metadata manifest](../assets/adding-status-informers.png)

> We haven't provided a status informer For Postgres since the customer can

Reviewer 2: ```suggestion
We haven't provided a status informer for Postgres since the customer can
```

choose to bring their own database. We'll add it in when we work with the
customer configuration in the next step of this lab.

Releasing Your Application
==========================

We're still not going to install our appliance, but let's release these

> changes to get into the habit or releasing changes quickly.

Reviewer 2: ```suggestion
changes to get into the habit of releasing changes quickly.
```

First we want to bump the version in our Helm chart. You could consider this
part optional since we didn't update the chart. I recommend you do it, since
different customer may install the application in different ways. It will be
much easier for your team if Helm installs using version `0.6.1` and appliance
installs using version `0.6.1` are using the same bits.

You can bump the chart version in your editor, or just run the following two
commands

```
yq -i '.version = "0.6.1"' slackernews/Chart.yaml
yq -i '.spec.chart.chartVersion = "0.6.1"' release/slackernews-chart.yaml
```

> You need to edit both files to assure that the appliance refers to the current

Reviewer 2: ```suggestion
You need to edit both files to ensure that the appliance refers to the current
```

chart. You also need to repackage you're Helm chart and make sure your old
package isn't still in the release directory.

```
helm package -u slackernews -d release
rm release/slackernews-0.6.0.tgz
```

Create a new release, and promote it directly to the "Unstable" channel.

```

> replicated release create --promote Unstable --chart ${HOME_DIR}/release/slackernews-0.6.1.tgz --version 0.6.1 \

Reviewer 2: Isn't this a Helm-only release?

Reviewer 2: This whole command doesn't work for me, and I can't proceed. Looks like `HOME_DIR` is empty, so it tries to find the chart at `/release/slackernews-0.6.1.tgz`. I don't know how that variable worked in the previous step though.
```
replicant@shell:~$ helm package -u slackernews -d release                                                                                                                        
Pulled: registry.replicated.com/library/replicated:1.0.0-beta.26
Digest: sha256:221b4d6ebb329e5d7e2d0bed4dafd59a130c77a58b8cefa24e6161e035ebb53d
Successfully packaged chart and saved it to: release/slackernews-0.6.1.tgz
replicant@shell:~$ replicated release create --promote Unstable --chart ${HOME_DIR}/release/slackernews-0.6.1.tgz --version 0.6.1 \
  --release-notes "Configures appliance branding, status, and dashboard links" \
  --app ${REPLICATED_APP} --token ${REPLICATED_API_TOKEN}
You are creating a release that will only be installable with the helm CLI.
For more information, see 
https://docs.replicated.com/vendor/helm-install#about-helm-installations-with-replicated

  _ Reading chart from /release/slackernews-0.6.1.tgz _  
ERROR: open /release/slackernews-0.6.1.tgz: no such file or directory
```

Author: Copy/paste error on my part. I copied it out of the script that I can test for the lab and didn't fix that up. You can set $HOME_DIR to $HOME to workaround but I'll fix.
